File: BackUpForABook/UIControls/TRClasses/TRRowAxis.m
//
//  TRRowAxis.m
//  ATreeRingMap
//
//  Created by Hui Wang on 5/26/14.
//  Copyright (c) 2014 AnnieLyticx. All rights reserved.
//

#import "TRRowAxis.h"

@interface TRRowAxis (PrivateMethods)

// paint middle slider bar
// start, end - defines the moving range for the slider
// angle - defines the line on which the slider is moving
-(void)paintMiddleSliderBar:(CGContextRef)ctx
                startRadius:(float)start
                  endRadius:(float)end
                      angle:(float)angle
                     center:(CGPoint)origin;

// paint the slider bar
-(void)paintMiddleBar:(CGContextRef)ctx
                angle:(float)angle
             position:(CGPoint)position;

// paint the labels on the left and right edges
// start, end - start and end of visual space
// angle - angle of the line where lables are drawn onto
// origin - the center of tree ring map
-(void)paintTickLabels_Angle:(CGContextRef)ctx
               startPosition:(float)start
                 endPosition:(float)end
                       angle:(float)angle
                      center:(CGPoint)origin
                 frameHeight:(int)height;

// paint labels on the left and right edges with 0 degree or shift labels up and down
-(void)paintTickLabels_Level:(CGContextRef)ctx
               startPosition:(float)start
                 endPosition:(float)end
                       angle:(float)angle
                      center:(CGPoint)origin
                 frameHeight:(int)height;

// paint label zero degree
-(void)paintLabel_0:(CGContextRef)ctx
              label:(TRLabel *)label
             string:(NSMutableAttributedString *)attString
        containRect:(CGRect) rect;

// paint ticks
-(void)paintTicks:(CGContextRef)ctx
           offset:(float)offset
             angle:(float)angle
           center:(CGPoint)origin;

@end

@implementation TRRowAxis

-(id)init
{
    self = [super init];
    
    // !!! self.tickLableDirection has two possible values: TickLableOrientation_0 and TickLableOrientation_angle
    self.tickLableDirection = TickLableOrientation_Angle;
    
    // show label
    self.showLable = TRUE;
    self.showTicks = TRUE;
    
    // slider bar
    _showBar = TRUE;
    _slider = [[DSSlider alloc] init];
    
    // data set for fishy eye effect
    super.eyeRadius = 50;
    super.eyeTicks = 3;
    super.eyePosition = (lastTickValue + firstTickValue) * .5;
    super.fishyEye = TRUE;
    
    return self;
}

// paint axis
// paint axis
// start, end - range of visual space
// angle - axis direction
// tickoffset - tick offset away form the axis 0 ground
// origin - center point of ring
-(void)paint:(CGContextRef)ctx
startPosition:(float)start
 endPosition:(float)end
       angle:(float)angle
      offset:(float)offset
      center:(CGPoint)origin
 frameHeight:(int)height
{
    if (self.showLable == TRUE)
    {
        // paint the label
        // !!! self.tickLableDirection has two possible values: TickLableOrientation_0 and TickLableOrientation_Angle
        if (self.tickLableDirection == TickLableOrientation_Angle)
        {
            // automatic angle orientation
            [self paintTickLabels_Angle:ctx startPosition:start endPosition:end angle:angle center:origin frameHeight:height];
        }else
        {
            // 0 degree or shifting up and down
            [self paintTickLabels_Level:ctx startPosition:start endPosition:end angle:angle center:origin frameHeight:height];
        }
    }
    
    if (self.showLable == TRUE && self.showTicks)
    {
        [self paintTicks:ctx offset:offset angle:angle center:origin];
    }
    
    return ;
}

// show the slider bar
-(void)paintSliderBar:(CGContextRef)ctx
        startPosition:(float)start
          endPosition:(float)end
                angle:(float)angle
               center:(CGPoint)origin
{
    if (_showBar == TRUE && self.fishEyeNeeded)
    {
        [self paintMiddleSliderBar:ctx startRadius:start endRadius:end angle:angle center:origin];
    }
}

// paint middle slider bar
// start, end - defines the moving range for the slider
// angle - defines the line on which the slider is moving
-(void)paintMiddleSliderBar:(CGContextRef)ctx
                startRadius:(float)start
                  endRadius:(float)end
                      angle:(float)angle
                     center:(CGPoint)origin
{
    // make sure barPostion is within half bar size range
    float halfBarSize = _slider.length * .5 + 8;
    
    if (super.eyePosition < (start + halfBarSize))
    {
        super.eyePosition = start + halfBarSize;
    }
    
    if (super.eyePosition > (end - halfBarSize))
    {
        super.eyePosition = end - halfBarSize;
    }
    
    CGPoint centerPosition = CGPointMake(origin.x + super.eyePosition * cosf(DEGREES_TO_RADIANS(angle)), origin.y - super.eyePosition * sinf(DEGREES_TO_RADIANS(angle)));
    
    [self paintMiddleBar:ctx angle:angle position:centerPosition];
}

// paint middle slider bar
-(void)paintMiddleBar:(CGContextRef)ctx
                angle:(float)angle
             position:(CGPoint)position
{
    // save current context
    CGContextSaveGState(ctx);
    
    CGContextTranslateCTM(ctx, position.x, position.y);
    CGContextRotateCTM(ctx, DEGREES_TO_RADIANS(-angle));
    
    // draw the arrow
    [_slider paint:ctx];
    
    // restore the context
    CGContextRestoreGState(ctx);
}

// paint the labels on the left and right edges
// start, end - start and end of visual space
// angle - angle of the line where lables are drawn onto
// isLeft - true if the left side axis
// origin - the center of tree ring map
-(void)paintTickLabels_Angle:(CGContextRef)ctx
               startPosition:(float)start
                 endPosition:(float)end
                       angle:(float)angle
                      center:(CGPoint)origin
                 frameHeight:(int)height
{
    int i;
    float rPosition;
    CGPoint labelPosition;
    float labelHeight, lableWidth;
    float labelSpace = .0;
    float labelAngle = .0;  // computed based on the avalible space
    CGRect txtSize;
    
    if (numberOfTicks == 0)
    {
        return;
    }
    
    // alignment
    NSMutableParagraphStyle *paragraphStyle = [[NSMutableParagraphStyle alloc] init];
    paragraphStyle.alignment = NSTextAlignmentLeft;
    
    // save current context
    CGContextSaveGState(ctx);
 
    // inital values.  will be reassigned dynamicly
    labelHeight = 40.0;
    lableWidth = 100.0;
    
    NSMutableAttributedString *sampleString = [[NSMutableAttributedString alloc] initWithString:[self tick:0].label.shortString attributes:[self tick:0].label.attrDictionary];
    txtSize = [sampleString boundingRectWithSize:CGSizeMake(CGFLOAT_MAX, CGFLOAT_MAX) options:0 context:nil];

    // labelAngle based on visual between lables
    if (numberOfTicks == 1)
    {
        labelAngle = .0;
    }else
    {
        labelAngle = 180 * atan(txtSize.size.height / ([[self tick:1] viewOffset] - [[self tick:0] viewOffset])) / 3.14;
    }
    
    // going over all axis tick text
    for (i = 0; i < numberOfTicks; i++)
    {
        // ith TRAxisTick object
        TRAxisTick *tick = [self tick:i];
        
        if (tick == nil)
        {
            continue;
        }
        
        // radius position of this lable
        rPosition = tick.viewOffset;
        
        // create attributed string for drawing
        NSMutableAttributedString *attString = [[NSMutableAttributedString alloc] initWithString:tick.label.shortString attributes:tick.label.attrDictionary];
        
        // check to see if we have enough space for displaying the label
        // get string label size
        txtSize = [attString boundingRectWithSize:CGSizeMake(CGFLOAT_MAX, CGFLOAT_MAX) options:0 context:nil];
        
        lableWidth = txtSize.size.width;
        labelHeight = txtSize.size.height;
        
        // adjust the label length based on what we have at this position
        labelSpace =  lableWidth;
        /* No need to make adjustment here.  let caller make decision on string length
        if (labelSpace > self.labelMargin * self.pointsPerFontSize)
        {
            labelSpace = self.labelMargin * self.pointsPerFontSize;
        }
        */
        
        // labe head poition
        labelPosition.x = origin.x + rPosition * cosf(DEGREES_TO_RADIANS(angle));
        labelPosition.y = origin.y - rPosition * sinf(DEGREES_TO_RADIANS(angle)) + tick.viewSpace;
        
        // draw this label
        CGRect rect = CGRectMake(.0, labelHeight, labelSpace, labelHeight + 2);
        
        CGContextTranslateCTM(ctx, labelPosition.x, labelPosition.y);
        CGContextRotateCTM(ctx, -DEGREES_TO_RADIANS(labelAngle));
        
        [self paintLabel_0:ctx label:tick.label string:attString containRect:rect];
        
        CGContextRotateCTM(ctx, DEGREES_TO_RADIANS(labelAngle));
        CGContextTranslateCTM(ctx, -labelPosition.x, -labelPosition.y);
    }
    
    // restore the context
    CGContextRestoreGState(ctx);
}

// paint labels on the left and right edges with 0 degree or shift labels up and down
-(void)paintTickLabels_Level:(CGContextRef)ctx
               startPosition:(float)start
                 endPosition:(float)end
                       angle:(float)angle
                      center:(CGPoint)origin
                 frameHeight:(int)height
{
    int i;
    float rPosition;
    CGPoint labelPosition;
    float labelHeight, lableWidth;
    float labelSpace = .0;
    float labelShift = .0;
    CGRect txtSize;
    
    if (numberOfTicks == 0)
    {
        return;
    }
    
    // alignment
    NSMutableParagraphStyle *paragraphStyle = [[NSMutableParagraphStyle alloc] init];
    paragraphStyle.alignment = NSTextAlignmentLeft;
    
    // save current context
    CGContextSaveGState(ctx);
    
    // inital values.  will be reassigned dynamicly
    labelHeight = 40.0;
    lableWidth = 100.0;
    
    NSMutableAttributedString *sampleString = [[NSMutableAttributedString alloc] initWithString:[self tick:0].label.shortString attributes:[self tick:0].label.attrDictionary];
    txtSize = [sampleString boundingRectWithSize:CGSizeMake(CGFLOAT_MAX, CGFLOAT_MAX) options:0 context:nil];
    
    // labelAngle based on visual between lables
    if (numberOfTicks == 1)
    {
        labelShift = .0;
    }else
    {
        labelShift = txtSize.size.height + 2;
    }
    
    // going over all axis tick text
    for (i = 0; i < numberOfTicks; i++)
    {
        // ith TRAxisTick object
        TRAxisTick *tick = [self tick:i];
        
        if (tick == nil)
        {
            continue;
        }
        
        // radius position of this lable
        rPosition = tick.viewOffset;
        
        // create attributed string for drawing
        NSMutableAttributedString *attString = [[NSMutableAttributedString alloc] initWithString:tick.label.shortString attributes:tick.label.attrDictionary];
        
        // check to see if we have enough space for displaying the label
        // get string label size
        txtSize = [attString boundingRectWithSize:CGSizeMake(CGFLOAT_MAX, CGFLOAT_MAX) options:0 context:nil];
        
        lableWidth = txtSize.size.width;
        labelHeight = txtSize.size.height;
        
        // adjust the label length based on what we have at this position
        labelSpace =  lableWidth;
        /* No need to make adjustment here.  let caller make decision on string length
        if (labelSpace > self.labelMargin * self.pointsPerFontSize)
        {
            labelSpace = self.labelMargin * self.pointsPerFontSize;
        }
        */
        
        // labe head poition
        labelPosition.x = origin.x + rPosition * cosf(DEGREES_TO_RADIANS(angle));
        labelPosition.y = origin.y - rPosition * sinf(DEGREES_TO_RADIANS(angle)) + tick.viewSpace;
        
        // make adjustment along x direction
        labelPosition.x -= .5 * lableWidth;
        
        // make adjustment along y direction
        if (i % 2 != 0)
        {
            labelPosition.y += labelShift;
        }
        
        // draw this label
        CGRect rect = CGRectMake(.0, labelHeight, labelSpace, labelHeight + 2);
        
        CGContextTranslateCTM(ctx, labelPosition.x, labelPosition.y);
        
        [self paintLabel_0:ctx label:tick.label string:attString containRect:rect];
        
        CGContextTranslateCTM(ctx, -labelPosition.x, -labelPosition.y);
    }
    
    // restore the context
    CGContextRestoreGState(ctx);
    return ;
}

// paint given text label at given rect position
// labe - label object that has string and visual drawing information
// rect - the rect area for string drawing
-(void)paintLabel_0:(CGContextRef)ctx
              label:(TRLabel *)label
             string:(NSMutableAttributedString *)attString
        containRect:(CGRect) rect
{
    // (1) Under line for the first letter?
    if (label.underLine == TRUE)
    {
        [attString addAttribute:NSUnderlineStyleAttributeName value:[NSNumber numberWithInt:2] range:NSMakeRange(0, 1)];
    }
    // (2) exceeds the maximum number of letters?
    if ([attString length] >= label.maxNumberOfFullSizeLetters && label.attrRemainderDictionary != nil)
    {
        [attString addAttributes:label.attrRemainderDictionary range:NSMakeRange(label.maxNumberOfFullSizeLetters, [attString length] - label.maxNumberOfFullSizeLetters)];
    }
    // (3) draw at labelPosition
    CGMutablePathRef path = CGPathCreateMutable();
    CGPathAddRect(path, NULL, rect);
    
    CTFramesetterRef framesetter = CTFramesetterCreateWithAttributedString((CFAttributedStringRef)attString);
    CTFrameRef frame = CTFramesetterCreateFrame(framesetter, CFRangeMake(0, [attString length]), path, NULL);
    
    if (label.blurRadius > .0)
    {
        CGContextSetShadowWithColor(ctx, label.blurSize, label.blurRadius, label.blurColor.CGColor);
    }
    
    // paint the label
    // save current context
    CGContextSaveGState(ctx);
    CGContextSetTextMatrix(ctx, CGAffineTransformMakeScale(1.0f, -1.0f));
    
    CTFrameDraw(frame, ctx);
    
    // restore the context
    CGContextRestoreGState(ctx);
    
    // release the objects
    CFRelease(frame);
    CFRelease(framesetter);
    CFRelease(path);
}

// paint ticks
-(void)paintTicks:(CGContextRef)ctx
           offset:(float)offset
            angle:(float)angle
           center:(CGPoint)origin
{
    int i;
    float rPosition;
    CGPoint axisPosition, labelPosition;
    
    // save current context
    CGContextSaveGState(ctx);
    
    CGContextSetStrokeColorWithColor(ctx, self.tickColor.CGColor);
    CGContextSetLineWidth(ctx, self.tickSize);
    
    // going over all axis tick text
    CGContextBeginPath(ctx);
    for (i = 0; i < numberOfTicks; i++)
    {
        // ith TRAxisTick object
        TRAxisTick *tick = [self tick:i];
        
        if (tick == nil)
        {
            continue;
        }
        
        // radius position of this lable
        rPosition = tick.viewOffset;
        
        // axis position
        axisPosition.x = origin.x + rPosition * cosf(DEGREES_TO_RADIANS(angle));
        axisPosition.y = origin.y - rPosition * sinf(DEGREES_TO_RADIANS(angle)) + offset;
        
        // labe head poition
        labelPosition.x = axisPosition.x;
        labelPosition.y = axisPosition.y + self.height * self.pointsPerFontSize;
        
        // make adjustment along y direction
        if (i % 2 != 0)
        {
            labelPosition.y = axisPosition.y + (self.height + 3.0) * self.pointsPerFontSize;
        }
        
        // draw line from axisPosition to labelPosition
        CGContextMoveToPoint(ctx, axisPosition.x, axisPosition.y);
        CGContextAddLineToPoint(ctx, labelPosition.x, labelPosition.y);
        
        CGContextStrokePath(ctx);
     }
    
    // restore the context
    CGContextRestoreGState(ctx);

    return ;
}

// hit test.  Potential hit objects: slider and and axis lable
-(HitObj)HitTest:(CGPoint)atPoint
           angle:(float)angle
         barSize:(float)size
          center:(CGPoint)origin
{
    HitObj hit;
    
    hit.hitObject = TRObjs_None;

    // convert atPoint to position in (angle, radius) coordinate position first
    CGFloat r = hypotf(atPoint.x - origin.x, atPoint.y - origin.y);
    float a = 180.0 * atan(-(atPoint.y - origin.y) / (atPoint.x - origin.x)) / 3.14;
    
    if ((atPoint.x - origin.x) < 0)
    {
        a += 180;
    }

    // (1) hit slider bar?
    bool visibleBar = (_showBar == TRUE && self.fishEyeNeeded);
    float barHalfWidthAngle = 180.0 * _slider.height / (3.14 * super.eyePosition);
    if (visibleBar &&
        (fabs(a - angle) <= barHalfWidthAngle || fabs(a - angle - 360.0) <= barHalfWidthAngle) &&
        fabsf(super.eyePosition - r) <= _slider.length * .5)
    {
        hit.hitObject = TRObjs_RowBar;
        return hit;
    }
    
    
    // (2) hit axis area?
    if (r > super.firstTickValue && r < super.lastTickValue)
    {
        float angleHaflRange = 180.0 * size / (3.14 * r);
        if (fabs(a - angle) <= angleHaflRange || fabs(a - angle - 360.0) <= angleHaflRange)
        {
            hit.hitObject = TRObjs_Row_Axis;
            return hit;
        }
    }
    
     // (3) hit axis lables?
    
    return hit;
}

@end

File: BackUpForABook/UIControls/TRClasses_swift/TRRowAxis.swift
//
//  TRRowAxis.swift
//  TreeRingMap_Swift
//
//  Created by dingf on 17/3/16.
//  Copyright © 2017年 MH.dingf. All rights reserved.
//

import UIKit

class TRRowAxis: TRAxis {
    var showBar: Bool?
    var slider: DSSlider?

    override init() {
        super.init()
        tickLabelDirection = .angle
        showTicks = true
        showLabel = true
        showBar = true
        slider = DSSlider.init()
        
        eyeRadius = 50
        eyeTicks = 3
        eyePosition = (lastTickValue! + firstTickValue!) * 0.5
        fishyEye = true
    }
    
    
    // paint 
    func paint(_ ctx: CGContext,start: Float,end: Float,angle: Float,offset: Float,origin: CGPoint,height: Int){
        if showLabel == true{
            if tickLabelDirection == .angle{
                // automatic angle orientation

                
            }else{
                // 0 degree or shifting up and down
                paintTickLabels_Level(ctx, start: start, end: end, angle: angle, origin: origin, height: height)
            }
            
            if showLabel == true && showTicks == true{
                paintTicks(ctx, offset: offset, angle: angle, origin: origin)
            }
        }
    }
    
    // paint ticks : small lines
    func paintTicks(_ ctx: CGContext, offset: Float, angle: Float, origin: CGPoint){
        ctx.saveGState()
        
        ctx.setStrokeColor(tickColor!.cgColor)
        ctx.setLineWidth(CGFloat(tickSize!))
        
        // go over all axis tick text
        ctx.beginPath()
        
        for i in 0..<numberOfTicks!{
            let tick = ticks[i]
            let rPosition = tick.viewOffset
            
            // axis position 
            let axisPosition = CGPoint(x: origin.x + CGFloat(rPosition! * cosf(DEGREETORADIANS(degree: angle))), y: origin.y - CGFloat(rPosition! * sinf(DEGREETORADIANS(degree: angle))) + CGFloat(offset))
            
            // label head position 
            var labelPosition = CGPoint(x: axisPosition.x, y: axisPosition.y + CGFloat(height! * pointsPerFontSize!))
            
            // make adjustment along y direction 
            if i % 2 != 0 {
                labelPosition.y = axisPosition.y + CGFloat((height! + 3.0) * pointsPerFontSize!)
            }
            
            ctx.move(to: axisPosition)
            ctx.addLine(to: labelPosition)
            ctx.strokePath()
        }
        ctx.restoreGState()
    }
    
    func paintTickLabels_Level(_ ctx: CGContext, start: Float, end: Float, angle: Float, origin: CGPoint, height: Int){
        var rPosition :Float
        var labelPosition = CGPoint()
        var labelHeight: Float
        var labelWidth: Float
        
        var labelSpace: Float = 0
        var labelShift: Float = 0
        var txtSize: CGRect
        
        if numberOfTicks == 0 {return}
        
        // alignment
        let paragarphStyle = NSMutableParagraphStyle()
        paragarphStyle.alignment = .left
        
        // save current context
        ctx.saveGState()
        
        labelHeight = 40.0
        labelWidth = 100.0
        
        let sampleString = NSMutableAttributedString.init(string: ticks[0].label!.shortString!, attributes: ticks[0].label!.attrDictionary)
        txtSize = sampleString.boundingRect(with: CGSize.init(width: CGFloat(FLT_MAX), height: CGFloat(FLT_MAX)), options: NSStringDrawingOptions.init(rawValue: 0), context: nil)
        
        if numberOfTicks == 1{
            labelShift = 0
        }else{
            labelShift = Float(txtSize.size.height) + 2
        }
        
        // go over all axis tick text
        for i in 0..<numberOfTicks!{
            let tick = ticks[i]
            
            // draw this tick label at aPosition angle position
            rPosition = tick.viewOffset!
            
            // create attributed string for drawing
            let attString = NSMutableAttributedString.init(string: tick.label!.shortString!, attributes: tick.label!.attrDictionary)
            
            // check to see if we have enough space for displaying the label : get string label size
            txtSize = attString.boundingRect(with: CGSize.init(width: CGFloat(FLT_MAX), height: CGFloat(FLT_MAX)), options: NSStringDrawingOptions.init(rawValue: 0), context: nil)
            labelWidth = Float(txtSize.size.width)
            labelHeight = Float(txtSize.size.height)
            
            // adjust the label length based on what we have at this position
            labelSpace = labelWidth
            
            // label head position
            labelPosition.x = origin.x + CGFloat(rPosition * cosf(DEGREETORADIANS(degree: angle)))
            labelPosition.y = origin.y - CGFloat(rPosition * sinf(DEGREETORADIANS(degree: angle))) + CGFloat(tick.viewSpace!)
            
            
            labelPosition.x -= CGFloat(0.5 * labelWidth)
            
            if (i % 2) != 0{
                labelPosition.y += CGFloat(labelShift)
            }
            
            
            let rect = CGRect(x: 0, y: CGFloat(labelHeight), width: CGFloat(labelSpace), height: CGFloat(labelHeight + 2))
            
            ctx.translateBy(x: labelPosition.x, y: labelPosition.y)
            paintLabel_0(ctx, label: tick.label!, attString: attString, rect: rect)
            ctx.translateBy(x: -labelPosition.x, y: -labelPosition.y)
        }
        
        //
        ctx.restoreGState()
    }
    
    
    // hit test. Potential hit objects: slider and axis label
    func hitTest(_ atPoint: CGPoint, angle: Float, size: Float, origin: CGPoint) -> HitObj{
        var hit = HitObj()
        hit.hitObject = TRObjs.none
        
        // convert atPoint to position in coordinate
        let r = hypotf(Float(atPoint.x - origin.x), Float(atPoint.y - origin.y))
        var a = 180.0 * atan(Float(-(atPoint.y - origin.y) / (atPoint.x - origin.x))) / 3.14
        if (atPoint.x - origin.x) < 0{
            a += 180
        }
        
        // 1. hit slider bar?
        let visibleBar = (showBar == true && fishEyeNeeded == true)
        let barHalfWidthAngle = 180.0 * slider!.length! / (3.14 * eyePosition!)
        if (visibleBar == true) && (fabs(a - angle) <= barHalfWidthAngle) || (fabs(a - angle - 360.0) <= barHalfWidthAngle) && (fabsf(eyePosition! - r) <= slider!.height! * 0.5){
            hit.hitObject = TRObjs.rowBar
            return hit
        }
        
        // 2. hit axis area?
        if r > firstTickValue! && r < lastTickValue!{
            let angleHalfRange = 180.0 * size / (3.14 * r)
            if (fabs(a - angle) <= angleHalfRange || fabs(a - angle - 360.0) <= angleHalfRange){
                hit.hitObject = TRObjs.row_axis
                return hit
            }
        }
        
        // 3. hit axis labels?
        
        return hit
    }
    
    
    func paintTickLabels_Angle(_ ctx: CGContext,start: Float,end: Float,angle: Float,origin:CGPoint,height: Int){
        if numberOfTicks == 0{return}
        //  alignment
        let paragraphStyle = NSMutableParagraphStyle()
        paragraphStyle.alignment = .left
        ctx.saveGState()
        
        let sampleString = NSMutableAttributedString(string:ticks[0].label!.shortString!, attributes: ticks[0].label!.attrDictionary)
        let txtSize = sampleString.boundingRect(with: CGSize.init(width: CGFloat.greatestFiniteMagnitude, height: CGFloat.greatestFiniteMagnitude), options: NSStringDrawingOptions.init(rawValue: 0), context: nil)
        var labelAngle: Float
        // labelangle based on visual between labels
        if numberOfTicks == 1{
            labelAngle = 0.0
        }else{
            let tick1 = ticks[1]
            let tick0 = ticks[0]
            labelAngle = 180 * atan(Float(txtSize.size.height) / (tick1.viewOffset!) - tick0.viewOffset! / 3.14)
        }
        // go over all axis tick text
        for i in 0..<numberOfTicks!{
            let tick = ticks[i]
            // radius position of this label
            let rPosition = tick.viewOffset
            // create attributed string for drawing
            let attString = NSMutableAttributedString.init(string: tick.label!.shortString!, attributes: tick.label!.attrDictionary)
            let txtSize = attString.boundingRect(with: CGSize.init(width: CGFloat(FLT_MAX), height: CGFloat(FLT_MAX)), options: NSStringDrawingOptions.init(rawValue: 0), context: nil)
            let labelWidth = txtSize.size.width
            let labelHeight = txtSize.size.height
            let labelSpace = labelWidth
            let labelPosition = CGPoint(x: origin.x + CGFloat(rPosition! * cosf(DEGREETORADIANS(degree: angle))), y: origin.y - CGFloat(rPosition! * sinf(DEGREETORADIANS(degree: angle)) + tick.viewSpace!))
            // draw this label
            let rect = CGRect(x: 0.0, y: labelHeight, width: labelSpace, height: labelHeight + 2)
            ctx.ctm.translatedBy(x: labelPosition.x, y: labelPosition.y)
            ctx.ctm.rotated(by: -(CGFloat)(DEGREETORADIANS(degree: labelAngle)))
            paintLabel_0(ctx, label: tick.label!, attString: attString, rect: rect)
            ctx.ctm.rotated(by: CGFloat(DEGREETORADIANS(degree: labelAngle)))
            ctx.ctm.translatedBy(x: -labelPosition.x, y: -labelPosition.y)
        }
        ctx.restoreGState()
    }
    
    func paintLabel_0(_ ctx: CGContext, label: TRLabel, attString: NSMutableAttributedString, rect: CGRect){
        // under line for the first letter
        if label.underLine == true{
            attString.addAttributes([NSUnderlineStyleAttributeName : NSNumber.init(value: 2)], range: NSRange.init(location: 0, length: 1))
        }
        // exceed the maximum number of letters?
        if attString.length >= label.maxNumberOfFullSizeLetters! && label.attrRemainderDictionary != nil {
            attString.addAttributes(label.attrRemainderDictionary!, range: NSRange.init(location: label.maxNumberOfFullSizeLetters!, length: attString.length - label.maxNumberOfFullSizeLetters!))
        }
        // draw at label position
        let path = CGMutablePath()
        path.addRect(rect)
        let framesetter = CTFramesetterCreateWithAttributedString(attString)
        let frame = CTFramesetterCreateFrame(framesetter, CFRangeMake(0, attString.length), path, nil)
        if label.blurRadius! > Float(0.0){
            ctx.setShadow(offset: label.blurSize!, blur: CGFloat(label.blurRadius!), color: label.blurColor!.cgColor)
        }
        ctx.saveGState()
        ctx.textMatrix = CGAffineTransform(scaleX: 1.0, y: -1.0)
        CTFrameDraw(frame, ctx)
        ctx.restoreGState()
    }
}


